pkg/component/observability/opentelemetry/operator/opentelemetryoperator.go

```diff
@@ -0,0 +1,408 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package operator
+
+import (
+	"context"
+	"time"
+
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	vpaautoscalingv1 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1"
+	"k8s.io/utils/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/component"
+	"github.com/gardener/gardener/pkg/resourcemanager/controller/garbagecollector/references"
+	"github.com/gardener/gardener/pkg/utils"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+)
+
+const (
+	// OperatorManagedResourceName is the name of the OpenTelemetry Operator managed resource.
+	OperatorManagedResourceName = name
+
+	name               = "opentelemetry-operator"
+	serviceAccountName = name
+	roleName           = name
+	clusterRoleName    = name
+)
+
+// Values keeps values for the OpenTelemetry Operator.
+type Values struct {
+	// Image is the image of the OpenTelemetry Operator.
+	Image string
+	// PriorityClassName is the name of the priority class of the OpenTelemetry Operator.
+	PriorityClassName string
+}
+
+type openTelemetryOperator struct {
+	client    client.Client
+	namespace string
+	values    Values
+}
+
+// NewOpenTelemetryOperator creates a new instance of OpenTelemetry Operator.
+func NewOpenTelemetryOperator(
+	client client.Client,
+	namespace string,
+	values Values,
+) component.DeployWaiter {
+	return &openTelemetryOperator{
+		client:    client,
+		namespace: namespace,
+		values:    values,
+	}
+}
+
+func (otel *openTelemetryOperator) Deploy(ctx context.Context) error {
+	var (
+		registry = managedresources.NewRegistry(kubernetes.SeedScheme, kubernetes.SeedCodec, kubernetes.SeedSerializer)
+
+		serviceAccount     = otel.serviceAccount()
+		clusterRole        = otel.clusterRole()
+		clusterRoleBinding = otel.clusterRoleBinding()
+		role               = otel.role()
+		roleBinding        = otel.roleBinding()
+		deployment         = otel.deployment()
+		vpa                = otel.vpa()
+	)
+
+	utilruntime.Must(references.InjectAnnotations(deployment))
+
+	serializedResources, err := registry.AddAllAndSerialize(
+		serviceAccount,
+		clusterRole,
+		clusterRoleBinding,
+		role,
+		roleBinding,
+		deployment,
+		vpa,
+	)
+	if err != nil {
+		return err
+	}
+
+	return managedresources.CreateForSeed(ctx, otel.client, otel.namespace, OperatorManagedResourceName, false, serializedResources)
+}
+
+func (otel *openTelemetryOperator) Destroy(ctx context.Context) error {
```

Comment:
// Destroy deletes the operator deployment.
```suggestion
func (o *openTelemetryOperator) Destroy(ctx context.Context) error {
```

Please, add doc strings to all other exported methods/funcs/vars.